### PR TITLE
Add aten::polar and its variants

### PR DESCRIPTION
**Pull request by yucai-intel**

### CODE REVIEW

**Comment by fengyuan14** [comment]
Looks we have better dtype coverage than expected?
image

**Comment by yucai-intel** [comment]
The data types are consistent with those supported by CUDA.

**Comment by fengyuan14** [comment]
Please skip the case, and add error log as comments

> The data types are consistent with those supported by CUDA. Screenshot 2024-07-22 152455

The error log is saying unexpected data type is supported in Polar backward. Do yo know which operator is invoked in backward? Can you have an investigation, if the backward op supports bfloat16, does not aligned with CUDA, you can skip the case and add error log and list the backward operator as comments.

**Comment by yucai-intel** [comment]
Polar's backward is calculated using complex(), which does not support bfloat16, so the test of this datatype is skipped.

**Comment by fengyuan14** [comment]
> o

If we align with CUDA, CUDA should fail as well?

**Comment by yucai-intel** [comment]
Cuda fails with same error:
RuntimeError: Expected both inputs to be Half, Float or Double tensors but got BFloat16 and BFloat16.

File: test/xpu/extended/run_test_with_skip.py [hunk 1]
```diff
@@ -154,6 +154,10 @@
     # Greatest relative difference: 0.00396728515625 at index (610,) (up to 0.001 allowed)
     "test_compare_cpu_hypot_xpu_bfloat16",
 
+    # RuntimeError: Expected both inputs to be Half, Float or Double tensors but got BFloat16 and BFloat16.
+    # Polar's backward is calculated using complex(), which does not support bfloat16. Cuda fails with same error.
```
Comment (fengyuan14) [inline test/xpu/extended/run_test_with_skip.py:158]:
```suggestion
    # Polar's backward is calculated using complex(), which does not support bfloat16. CUDA fails with same error.
```

File: src/ATen/native/xpu/TensorFactories.cpp [hunk 2]
```diff
@@ -151,6 +151,29 @@ Tensor& XPUNativeFunctions::complex_out(
...
+  return result;
+}
+
+Tensor XPUNativeFunctions::polar(const Tensor& abs, const Tensor& angle) {
+  complex_check_floating(abs, angle);
+  c10::TensorOptions options = abs.options();
+  options = options.dtype(toComplexType(abs.scalar_type()));
+  Tensor result = at::empty(0, options);
+  return at::polar_out(result, abs, angle);
+}
```
Comment (fengyuan14) [inline src/ATen/native/xpu/TensorFactories.cpp:175]:
```suggestion
```

File: yaml/xpu_functions.yaml [hunk 3]
```diff
@@ -268,6 +268,8 @@ supported:
   - eye.m_out
   - _efficientzerotensor
   - complex.out
+  - polar.out
+  - polar
```
Comment (fengyuan14) [inline yaml/xpu_functions.yaml:272]:
```suggestion
```

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**